### ISSUE TITLE
Add support for categories and/or tags (keywords)

### ISSUE THREAD

**Issue by t7ko** [vimwiki/vimwiki#85]
I'd like to implement this support.  But I'm new to vimwiki and don't undestand the core philosophy deep enough, so I need some guidance -- how, in general terms, this should be architected.

In particular, the key questions are:
- What should be the syntax?
- What would be the best moment to collect/update list of pages for a given Category/Keyword?

Crazy ideas:
- Add `[[@Category/the name here]]` anywhere in the document.  And then add on-write hook to append to that page.  And then the category page is simply another link.
  - Pros: no new syntax, looks more or less natural, and similar to Wiki.
  - The downside is: subdirectory is treated as a separate wiki, links would need to be put in there with relative path `../link name` to be able to reach upwards.
- Similar to previous: `[[@Category - Categ Name]]`.  Solves the downside above, but pollutes the namespace.

**Comment by t7ko** [comment]
@EinfachToll : I've started working on this feature.  I've got basic stuff ready -- updating metadata file on page save.  Would you like to pre-review it?  It would probably help me.  Alternatively, I can continue and complete entire feature -- but then, it will be more effort to re-do stuff if you find the approach to be wrong.

**Comment by EinfachToll** [comment]
Yes, if you e.g. put it in an extra branch in your private fork, I'll happily take a look.

**Comment by t7ko** [comment]
@EinfachToll : its under name 'tags' ([here](https://github.com/t7ko/vimwiki/tree/tags)).  Yesterday and today, I have implemented pretty much everything the basic version needs.  The only interesting feature that's missing now is search by multiple tags.  Not even sure if it's needed.

What's implemented:
* metadata file support (load, save)
* updating metadata on page save; per-wiki flag to enable/disable it
* tag completion when editing page content
* bulk re-generate metadata for all pages (command)
* docs
* syntax definitions (highlight)
* cmd to search by tag
* cmd to generate wiki page with tags and their "instances"

Notes:
* I'm using very simple format for metadata file (tab-separated list).  It's really easy, 30 lines of code total.  And still provides all that seems to be necessary.
* I'm using "tags" (not "categories") everywhere.

Let me know if you want me to create a pull request out of it.

**Comment by EinfachToll** [comment]
Ok, I tried it and I see that the new commands do something, at least ;)

Some conceptual remarks:
- currently, a tag < 5 lines below a header tags the header and every other tag tags the whole file. But what I really would like to have is that tags inside the text tags the line it is in. So that `:VimwikiGenerateTags mytag` creates links like `[[file_where_mytag_occurs#mytag]]`. When pressing enter, Vim jumps directly to the occurrence of mytag in the file. In other words, every tag is also an anchor. (But only if the tag doesn't sit < 5 lines below a header. In this case, the created link would be `[[file_where_mytag_occurs#header_with_mytag]]`. And if mytag appears, say, at the beginning of the file (above the first header), then the link is only `[[file_where_mytag_occurs]]`).
- Vim has powerful support for tags inbuild. So, if the .tags file format would be identical to Vim's tags file format, we would get many useful key combinations and commands dealing with tags for free. Maybe you can read through `:h tagsrch.txt` (personally, I don't have much time these days) and find out if it's possible.

Other issues:
- I have source code snippets in my wiki files, so I get many many false positives, that is, it shows me tags that are not meant to be tags (e.g. in my file about C++, in a line containing `std::getline(std::cin, bla)`, `:getline(std:` is considered a tag.) So we have to decide more precisely what a tag looks like. Maybe a tag must be preceded by space (or start of line) and followed by spaces (or end of line) and must not contain Ascii punctuation characters (except for - maybe).
- why does `:VimwikiSearchTags` search through all files instead of getting the location from the .tags file? Isn't this the whole point of tags?

**Comment by EinfachToll** [comment]
I invited you as a contributor. You can open a new branch 'tags' where you can push your stuff to.

**Comment by t7ko** [comment]
@EinfachToll 

Pushed to new branch `tags`.

Reworked to use CTAGS format, now Vim's built-in tags support can use `.tags` metadata file.

"Tags as Anchors" -- are yet to be worked on.

Issues with code snippets.  Options, in addition to those listed in your comment:
* Use some other delimiter, e.g. semicolon `;`.  Semicolon will definitely decrease amount of false positives.  *I'd prefer this option as it is the easiest one.*  Let me know what you think.
* Require that tags are always put on a separate line, which starts with `Tags: `, or `Keywords: `.

On the search command question.  My reasoning was:
1. Using `.tags` for search would be more complex to implement, and more error-prone, than full scan (which is a one-line command now).
2. Using full scan will probably not be noticeably slower than `.tags`.
3. Using full scan will always work correctly, even if `.tags` is outdated.

**Comment by t7ko** [comment]
> Isn't this the whole point of tags?

No, the .tags file was only needed for those things that cannot be easily done by direct files access/manipulation, namely:

1. Tags completion.
2. GenerateTags, which, in general, needs all info on all tags at once.
3. Multi-tags search -- if I ever come to implement it, as I don't think now that it's really needed.

**Comment by t7ko** [comment]
@EinfachToll -- "Tags as Anchors" -- implemented and pushed to `tags` branch.

Seems like last issue standing is false positives on C++ code with lots of colons.  Pls review my ideas above and let me know which way you'd like this to go.  I'd prefer semicolons approach.

**Comment by linuxcaffe** [comment]
personally, I like :tag1:tag2: format with colons a lot, partly because it is also used by several projects in my scope (orgmode and ledger-cli and..) it's visually clear and compact, and because I have no c++ code snippets to deal with. But you gotta do what you gotta do!

**Comment by t7ko** [comment]
@linuxcaffe -- thanks for inputs. :)

@EinfachToll -- could you please run a `:vimgrep` search on your wiki and see how many  false-positives you'd get with the following regex: `/\(^\|\s\):\([^:[:space:]]\+:\)\+\(\s\|$\)/` ?

(It implements your idea "left colon is first char in the line or preceded by space, right colon is last char in line or followed by a space".)

If you're ok with the false-positive rate of that regex, I could implement it.

BTW, another false positive I have encountered in my own wiki -- url followed by a colon, as in
> some text (http://some-url.com): some other text

(Which by the way highlights a bug -- vimwiki treats brace and colon which follow the url, as part of the url.)

**Comment by EinfachToll** [comment]
Uh, somehow Github didn't notify me of your latest comments here. Strange.

Regarding the search command: never mind, now that .tags has the ctags format, the user can use Vim's builtin :tag command. At the moment, that's not super nice, because `<BS>` doesn't work after jumping, and maybe other things break. I'm looking forward to investigate how Vimwiki and the tags feature can be married.

The tags syntax: I also like the `:tag1:tag2:` syntax and your regex did only produce two false tags, so I'm ok with it.

BTW: the highlight of the URL is not really a bug, because a URL can contain nearly every character, including () and :

**Comment by t7ko** [comment]
@EinfachToll -- did the fixes to prevent false positives.  Let me know if we need anything else here before letting it into `dev`.

**Comment by EinfachToll** [comment]
A small thing I noticed is that the auto completion starts only if you have already typed the beginning of a tag. It should be easy to change it so that the completion also works if you press ctrl+x ctrl+o directly behind a colon. Then, the popup menu of couse contains all tags there are.

**Comment by t7ko** [comment]
@EinfachToll - thanks for the catch, that was a bug.  Fixed now, pushed to `tags`.

**Comment by EinfachToll** [comment]
There was a problem with the auto completion. Sometimes, it didn't work, because, apparently, col('.') inside the completion function sometimes returns the start of the string to be completed, not the position where the cursor was when the user hits ctrl+x ctrl+o. I think I fixed it.
So, more testing … :)

**Comment by t7ko** [comment]
@EinfachToll -- ok.  One question/idea -- I could merge dev into tags, to make testing easier, and to prevent conflicts during future backmerge of tags into dev.  (I'm keeping my personal dev branch updated with vimwiki/dev and vimwiki/tags, and today I have encountered the first merge conflict, so I figure it will be easier to resolve them in future if we take actions now.)

**Comment by EinfachToll** [comment]
Yes, you can do so.

BTW, as you might have seen, I have changed all occurrences of == and friends to ==? or ==#, because recently I stumbled over this document which explains why: http://google-styleguide.googlecode.com/svn/trunk/vimscriptguide.xml .
So, if you want to contribute to Vimwiki also in the future, which of course I would appreciate, maybe you want to take a look at it. I'm going to repair the regular expressions in the Vimwiki code as well the next days.

**Comment by ches** [comment]
> At the moment, that's not super nice, because <BS> doesn't work after jumping, and maybe other things break. I'm looking forward to investigate how Vimwiki and the tags feature can be married.

FWIW I [recently implemented][1] wrappers for all the standard Vim tag jumping mappings on the vim-erlang-tags plugin that behave—I believe—faithfully to the originals. The particulars of why the wrappers are needed are of course specific to Erlang and the custom ctags generation they're doing on that project, but the general approach seems to be working out fine so far. That might give a small head start to Vimwiki for hooking the `<BS>` jump history, etc. 

[1]: https://github.com/vim-erlang/vim-erlang-tags/pull/22

**Comment by t7ko** [comment]
@EinfachToll -- merged dev into tags.  Will definitely look at that style guide.  Sad thing is it won't open in browser :(  I'll need to check out the folder and try open locally...

**Comment by t7ko** [comment]
@EinfachToll -- looked at the style guide.  Do you plan on applying all of these rules?

**Comment by EinfachToll** [comment]
No, I would only regard it as a good inspiration. Expecially the points below "Portability" are worth remembering. But the current coding style should not change, if there is no good reason.

**Comment by EinfachToll** [comment]
FYI, I moved all tags-related functions to an extra file, so that base.vim doesn't turn into a big dump of unrelated code. Also, I renamed some functions to s:foo(). That makes the interface of a file clearly visible, because functions of the form s:foo() can't be called from outside the file.

And I found that currently tags are not found if they appear in a file that is in a subdirectory. So, @t7ko, if you are bored, there is something to fix ;)

**Comment by t7ko** [comment]
@EinfachToll -- thanks for fixes!  About the subdirectory thing -- I need more info here.  What do you mean by "not found"?  How/when does it happen?  When saving a file?  When jumping to a tag?  When regenerating all tags?

**Comment by EinfachToll** [comment]
When I have a file `lala/lolo.wiki` in a subdirectory, and it contains a tag `:atag:`, then in the .tags file there is a line `atag    lolo.wiki       2;"     vimwiki:lolo\tlolo`. The information that the file is in a subdirectory is lost. So, when I type `:tag atag` it isn't found and when I do `:VimwikiGenerateTags` in index.wiki it produces a link `[[lolo]]` to a file which doesn't exist.

**Comment by t7ko** [comment]
Got it.  Just so you know, the chances of me getting to work on it in a nearest month are quite low (though not zero).  :(

**Comment by linuxcaffe** [comment]
@t7ko just so ya know, some of us are sitting at the edge of our chairs for this feature :-)

**Comment by t7ko** [comment]
@EinfachToll -- fixed the issue with tags in sub-directories.

@linuxcaffe -- you can check out `vimwiki/tags` branch -- it has the tags feature _and_ all the recent fixes in `dev` branch.

**Comment by t7ko** [comment]
@EinfachToll @linuxcaffe -- I have updated `vimwiki/tags` with recent changes from `dev` branch.  @EinfachToll, let me know if you want me to fix anything else in the branch?

**Comment by EinfachToll** [comment]
There are some things that must be done before releasing:
- [ ] make tags work in markdown and media syntax as well
- [ ] HTML conversion: a tag must have a label and a (unobstrusive) highlight
- [ ] VimwikiGenerateTags should replace old generated tags (I know, VimwikiGenerateLinks doesn't do it, but I think that GenerateTags will be used much more)
- [ ] improve the handling when using tags via Vim's inbuild tags commands. Especially that you can go back with backspace.

And then there are some not so important things I would like to have:
- [ ] when doing RebuildTags, only look in wiki files that are newer than the .tags file (I already started working on this)
- [ ] instead of including the line number in the .tags file, use a search command, so that you can jump not only to the right line, but also the right column
- [ ] maybe `<CR>` on a tag lets you cycle to all occurrences of a tag

Of course, all these points can be discussed and, of course, I don't ask you to implement everything. I hope that I have some free time next week.

**Comment by linuxcaffe** [comment]
Will the _ever_ be merged? I like the feature a lot, but I'm not sure if I should build on the dev branch or the tags branch. @EinfachToll and maybe some of those good things could be addressed as separate issues??

or is it just not ready?

**Comment by EinfachToll** [comment]
> or is it just not ready?

Exactly. You see, I'm short of time these days :/ But I promise to take care of it.

**Comment by stevenleeg** [comment]
Hey, any status on when this'll get merged? I'm super excited to add tags to my vimwiki workflow. 

Keep up the good work :+1: :+1:

**Comment by EinfachToll** [comment]
This is merged into the master branch for some weeks now. I just keep this issue open because of some little improvement plans in this comment: https://github.com/vimwiki/vimwiki/issues/85#issuecomment-99496695

**Comment by griffinqiu** [comment]
+10086

**Comment by jean** [comment]
Maybe the remaining work could be split out to separate issues .. that might make it easier for others to pick them up, or to pick them off one by one.

**Comment by tinmarino** [comment]
Closing for inactive, see the `tag` tag. There is a wonder tag generation and lookup using vim ctag primitives (and format).
